### DISCUSSION TITLE
how work custom font in avalonia 11.0.0 ?

### DISCUSSION THREAD

**Discussion by CTUzkc** [AvaloniaUI/Avalonia#12258] [Q&A]
i want to custom font to avalonia 11.0.0,because app threw exception "default font family name can not be null or empty" on ubuntu 16.04 x64.
previously successful one way to custom class ： "IFontManagerImpl" ,and add ttf file to resource in 10.x before.
now 11.0.0 ,IFontManagerImpl can not used.

**Comment by workgroupengineering** [comment]
https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Fonts.Inter

**Reply by CTUzkc** [reply]
thank you very much to provide some details of avalonia.fonts.inter?

**Reply by timunie** [reply]
it can be installed as a nuget package. It is there if you want your App to look the same everywhere.

**Reply by CTUzkc** [reply]
how work to use ttf(font file) in app by avalonia.fonts.inter?can you provide some details codes ? i am code rookie,thank you

**Reply by timunie** [reply]
Inter is special font, not meant to load custom font. Probably a missunderstanding. More about custom fonts in the docs: 

https://docs.avaloniaui.net/docs/next/guides/styles-and-resources/how-to-use-fonts
also #11822

**Reply by CTUzkc** [reply]
did you know other way to use custom cass derive this interface "IFontCollection" ,and use AppBuilder.ConfigureFonts() in Program.Main.
i want to use custom font (microsoft yh) in app,solve app run exception on ubuntu for overall situation.
thank you very much.

**Reply by CTUzkc** [reply]
i try this way "https://docs.avaloniaui.net/docs/next/guides/styles-and-resources/how-to-use-fonts"
can not solve that  exception,because i want to use chinese font in app run to ubuntu .
before use IFontManagerImpl ,now this interface not public,Looking for an alternative method.

**Reply by timunie** [reply]
Which exception please?

**Reply by CTUzkc** [reply]
exception "default font family name can not be null or empty" on ubuntu 16.04 x64.
vs 2022(chinese Lang) + avaloina 11.0.0 ,ubuntu(chinese Lang)

**Reply by CTUzkc** [reply]
https://github.com/CTUzkc/Imaage/blob/main/1.PNG
https://github.com/CTUzkc/Imaage/blob/main/2.PNG

sorry,I don't know how to upload pictures

**Reply by timunie** [reply]
The images are okay. How do you set the font?

**Reply by Gillibald** [reply]
DefaultFontFamily = new FontFamily("fonts:Inter")

**Reply by CTUzkc** [reply]
this is one way of old version(10.x) :
`public class CustomFontManagerImpl : IFontManagerImpl
    {
        private readonly Typeface[] _customTypefaces; 
        private readonly string _defaultFamilyName;        
        //Load font resources in the project, you can load multiple font resources
        private readonly Typeface _defaultTypeface = new Typeface("resm:AvaloniaApplication1.msyh#微软雅黑"); 

        public CustomFontManagerImpl()
        {
            _customTypefaces = new[] { _defaultTypeface };
            _defaultFamilyName = _defaultTypeface.FontFamily.FamilyNames.PrimaryFamilyName;
        }

        public string GetDefaultFontFamilyName()
        {
            return _defaultFamilyName;
        }

        public IEnumerable<string> GetInstalledFontFamilyNames(bool checkForUpdates = false)
        {
            return _customTypefaces.Select(x => x.FontFamily.Name);
        }

        private readonly string[] _bcp47 = { CultureInfo.CurrentCulture.ThreeLetterISOLanguageName, CultureInfo.CurrentCulture.TwoLetterISOLanguageName };

        public bool TryMatchCharacter(int codepoint, FontStyle fontStyle, FontWeight fontWeight, FontFamily fontFamily,CultureInfo culture, out Typeface typeface)
        {
            foreach (var customTypeface in _customTypefaces)
            {
                if (customTypeface.GlyphTypeface.GetGlyph((uint)codepoint) == 0)
                {
                    continue;
                }

                typeface = new Typeface(customTypeface.FontFamily.Name, fontStyle, fontWeight); return true;
            }
            var fallback = SKFontManager.Default.MatchCharacter(fontFamily?.Name, (SKFontStyleWeight)fontWeight,   SKFontStyleWidth.Normal, (SKFontStyleSlant)fontStyle, _bcp47, codepoint);

            typeface = new Typeface(fallback?.FamilyName ?? _defaultFamilyName, fontStyle, fontWeight); return true;
        }

        public IGlyphTypefaceImpl CreateGlyphTypeface(Typeface typeface)
        {
            SKTypeface skTypeface; switch (typeface.FontFamily.Name)
            {
                case FontFamily.DefaultFontFamilyName:
                case "微软雅黑":  //font family name
                    skTypeface = SKTypeface.FromFamilyName(_defaultTypeface.FontFamily.Name); break;
                default:
                    skTypeface = SKTypeface.FromFamilyName(typeface.FontFamily.Name,
                        (SKFontStyleWeight)typeface.Weight, SKFontStyleWidth.Normal, (SKFontStyleSlant)typeface.Style); break;
            }
            return new GlyphTypefaceImpl(skTypeface);
        }
    }`
and set ttf file in app to resources
and add this code:
` public override void RegisterServices()
        {
            AvaloniaLocator.CurrentMutable.Bind<IFontManagerImpl>().ToConstant(new CustomFontManagerImpl());
            base.RegisterServices();
        }` in app.axaml.cs
,that can be run on ubuntu.
but now 11.0.0 ,interface "IFontManagerImpl " can not be achieve,so my question is how work use "IFontManagerImpl " of alternative methods ?can you provide some detail code?

**Reply by CTUzkc** [reply]
`DefaultFontFamily = new FontFamily("fonts:Inter")`
can you provide some detail code? i see DefaultFontFamily can not set in ` AppBuilder ConfigureFonts(Action<FontManager> action)`

**Reply by Gillibald** [reply]
.With(new FontManagerOptions({ DefaultFontFamily = new FontFamily("fonts:Inter")) add this to your AppBuilder

**Reply by CTUzkc** [reply]
only this codes: `.With(new FontManagerOptions() { DefaultFamilyName = "Fonts:微软雅黑 or fonts:Inter" })` in avalonia 11.0.0,but app running exception "System.InvalidOperationException:Could not create glyphTypeface.”
Did I write something wrong？
Are you considering improvement Simpler to Use custom font file next version ?

**Reply by Gillibald** [reply]
You have to reference the Inter font package and call UseInterFont on your AppBuilder

**Reply by CTUzkc** [reply]
I Tried.Still can't solve that problem.This problem is somewhat complex,Maybe we still need to start from `AppBuilder.ConfigureFonts()` to solve it.Do you know the details of IFontCollection，for example “IFontCollection.TryMatchCharacter” and “IFontCollection.TryGetGlyphTypeface” ?
